stan/math/prim/fun/select.hpp

```diff
@@ -0,0 +1,176 @@
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_true, typename T_false,
+          typename ReturnT = promote_scalar_t<return_type_t<T_true, T_false>,
+                                              plain_type_t<T_true>>,
+          require_container_t<T_true>* = nullptr,
+          require_stan_scalar_t<T_false>* = nullptr>
+inline ReturnT select(const bool c, const T_true& y_true,
+                      const T_false& y_false) {
+  if (c) {
+    return y_true;
+  } else {
+    return apply_scalar_binary(
+        y_true, y_false,
+        [](const auto& y_true_inner, const auto& y_false_inner) {
+          return y_false_inner;
+        });
+  }
+}
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument.
+ *
+ * Overload for use when the 'true' return is a scalar and the 'false'
+ * return is a container
+ *
+ * If chosen, the scalar is returned as a container of the same size and
+ * plain type as the provided argument. Consequently, any Eigen expressions are
+ * evaluated.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_true, typename T_false,
+          typename ReturnT = promote_scalar_t<return_type_t<T_true, T_false>,
+                                              plain_type_t<T_false>>,
+          require_stan_scalar_t<T_true>* = nullptr,
+          require_container_t<T_false>* = nullptr>
+inline ReturnT select(const bool c, const T_true y_true,
+                      const T_false y_false) {
+  if (c) {
+    return apply_scalar_binary(
+        y_true, y_false,
+        [](const auto& y_true_inner, const auto& y_false_inner) {
+          return y_true_inner;
+        });
+  } else {
+    return y_false;
+  }
+}
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument. Overload for use with an Eigen
+ * object of booleans, and two scalars.
+ *
+ * The chosen scalar is returned as an Eigen object of the same dimension
+ * as the input Eigen argument
+ *
+ * @tparam T_bool type of Eigen boolean object
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Eigen object of boolean condition values.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_bool, typename T_true, typename T_false,
+          require_eigen_array_vt<std::is_integral, T_bool>* = nullptr,
+          require_all_stan_scalar_t<T_true, T_false>* = nullptr>
+inline auto select(const T_bool c, const T_true y_true, const T_false y_false) {
+  return c.unaryExpr([&](bool cond) { return cond ? y_true : y_false; }).eval();
```

Comment:
(optional) I personally prefer we just pull in the names of the things we want into a lambda instead of using &
```suggestion
  return c.unaryExpr([y_true, y_false](bool cond) { return cond ? y_true : y_false; }).eval();
```